### PR TITLE
Issue 1681 genetic algorithm

### LINKED ISSUE #1681: Optimization: Genetic Algorithm
---

Author Name: **Robert Braun** (@robbr48)
Original Redmine Issue: 1681, https://flumes.iei.liu.se/redmine/issues/1681
Original Date: 2018/02/08
Original Assignee: Robert Braun

---

None

### PR DESCRIPTION
**Pull request by robbr48**
Added a genetic optimization algorithm. Resolves #1681.

### CODE REVIEW

File: HopsanGUI/Dialogs/OptimizationDialog.cpp [hunk 1]
```diff
@@ -1954,6 +1997,23 @@ void OptimizationDialog::recreateCoreProgressBars()
             mpCoreProgressBarsLayout->addWidget(mCoreProgressBarPtrs.last(),0,1);
         }
         break;
+    case Ops::Genetic:    //Particle swarm
```
Comment (peterNordin) [inline HopsanGUI/Dialogs/OptimizationDialog.cpp]: Wrong comment

File: Ops/src/OpsWorkerGenetic.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+
+//! @brief Initialization function
+void WorkerGenetic::initialize()
+{
+    Worker::initialize();
+}
+
+
+//! @brief Executes a particle swarm algorithm. optParticleInit() must be called before this one.
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: wrong documentation

File: Ops/src/OpsWorkerGenetic.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            }
+        }
+
+
+
+        mpMessageHandler->objectivesChanged();
+
+        //Check convergence
+        if(checkForConvergence()) break;      //Use complex method, it's the same principle
+
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: I am not sure, but if we break here wont the stepCompleted message be one counter behind?

File: Ops/src/OpsWorkerGenetic.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+//! @brief Set mutation probability parameter
+void WorkerGenetic::setMutationProbability(double value)
+{
+    mMutationProbability = value;
+}
+
+
+//! @brief Auxiliary function for selecting parents for crossover
+void WorkerGenetic::selectParents() {
+    double bestValue = 1e100;
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: Use std::numeric_limits<double>::max()

File: Ops/src/OpsWorkerGenetic.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+
+//! @brief Auxiliary function for mutating children
+void WorkerGenetic::mutate()
+{
+    for(size_t i=0; i<mNumPoints; ++i) {
+        for(size_t j=0; j<mNumParameters; ++j) {
+            double doMutation = opsRand();
+            if(doMutation < mMutationProbability) {
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: Here the check is <, but above for crossover it was >, is that intentional?

File: Ops/src/OpsWorkerGenetic.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                double mean = mCandidatePoints[i][j];
+                double dev = (mParameterMax[j]-mParameterMin[j])/2.0;
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: Is a fabs() neede here? Cant have negative dev I assume

File: Scripts/HCOM/optTemplateGenetic.hcom [hunk 7]
```diff
@@ -0,0 +1,42 @@
+# Auto-generated HCOM script for particle swarm optimization algorithm
+
```
Comment (peterNordin) [inline Scripts/HCOM/optTemplateGenetic.hcom]: Copy paste error in comment

File: Ops/src/OpsWorkerGenetic.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    double bestValue = 1e100;
+    for(size_t i=0; i<mNumPoints; ++i) {
+        double value = mObjectives[i]*opsRand();
+        if(value < bestValue) {
+            mParent1 = i;
+            bestValue = value;
+        }
+    }
+
+    bestValue = 1e100;
```
Comment (peterNordin) [inline Ops/src/OpsWorkerGenetic.cpp]: Same here

File: Ops/include/OpsWorkerGenetic.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  void selectParents();
+  void crossOver();
+  void mutate();
+  double gaussian(double mean, double dev, double min, double max);
+
+  size_t mParent1,mParent2;
+  std::vector<double> mChild1, mChild2;
+
+  double mCrossoverProbability;
+  double mMutationProbability;
```
Comment (peterNordin) [inline Ops/include/OpsWorkerGenetic.h]: These two should br initialized with some resonable defaults I think. Or the algorithm should not run if they are not set. (That would actually be better). This is maybe a more general problem to solve..,

File: HopsanGUI/Dialogs/OptimizationDialog.cpp [hunk 10]
```diff
@@ -1098,6 +1113,25 @@ void OptimizationDialog::generateDifferentialEvolutionScript()
     mScript = templateCode;
 }
 
+void OptimizationDialog::generateGeneticScript()
+{
+    QFile templateFile(gpDesktopHandler->getExecPath()+"../Scripts/HCOM/optTemplateGenetic.hcom");
```
Comment (peterNordin) [inline HopsanGUI/Dialogs/OptimizationDialog.cpp]: Dont we have a getScripts path (or similar) not sure. Also no checking if file exists. (It should, but ......)

File: HopsanCLI/main.cpp [hunk 11]
```diff
@@ -422,6 +423,8 @@ int main(int argc, char *argv[])
             double C1 = 2;
             double C2 = 2;
             double vmax = 2;
+            double CP = 0.2;
+            double MP = 0.1;
```
Comment (peterNordin) [inline HopsanCLI/main.cpp]: These defaults are not the same as in the gui case

File: Ops/src/OpsWorkerGenetic.cpp (hunk 5 above)
Comment (robbr48) [reply Ops/src/OpsWorkerGenetic.cpp]: Yes, it is correct. When comparing crossover the conditions in the if statements are reversed, so it actually checks for not performing the  crossover.

File: HopsanGUI/Dialogs/OptimizationDialog.cpp (hunk 10 above)
Comment (robbr48) [reply HopsanGUI/Dialogs/OptimizationDialog.cpp]: No, the getScripts function returns the scripts folder under Documents. We want the one in the installation directory here.

**Pull request merged**